exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+	"go.opentelemetry.io/otel/sdk/metric"
+)
+
+type metricKey struct {
+	desc    *export.Descriptor
+	encoded string
+}
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	sync.RWMutex
+
+	labelEncoder export.LabelEncoder
+	handler      http.Handler
+	registerer   prometheus.Registerer
+	gatherer     prometheus.Gatherer
+	counters     counters
+	gauges       gauges
+	histograms   histograms
+}
+
+var _ export.Exporter = &Exporter{}
+var _ http.Handler = &Exporter{}
+var _ export.LabelEncoder = &Exporter{}
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registry is the prometheus registry that will be used as the default Registerer and
+	// Gatherer if these are not specified.
+	//
+	// If not set a new empty Registry is created.
+	Registry *prometheus.Registry
+
+	// Registerer is the prometheus registerer to register
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the system-default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// LabelEncoder is the label encoder that will be used to group
+	// and export metrics to Prometheus.
+	LabelEncoder export.LabelEncoder
```

Comment:
See comment below. I'm not sure we need to define an encoder.

Reply:
Removed this from options.